packages/app-mobile/components/NoteEditor/CodeMirror/RegionSpec.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * Describes an inline region's formatting. For example, an italicized region.
+ */
+
+import { Text as DocumentText, SelectionRange } from '@codemirror/state';
+
+export default class RegionSpec {
+	// Patterns to use when creating an instance of the tag.
+	// E.g.
+	//   templateStart=**
+	//   templateStop=**
+	//  would be used to create
+	//   **content here**
+	public templateStart: string;
+	public templateStop: string;
+
+	// Size of the buffer used to check for a match.
+	// Particularly useful for [startExp], [stopExp].
+	// Only used when the selection is empty.
+	protected startBuffSize: number;
+	protected stopBuffSize: number;
+
+	// Regular expressions for matching possible starting/stopping
+	// regions.
+	protected startExp?: RegExp;
+	protected stopExp?: RegExp;
+
+	public constructor({
+		templateStart, templateStop,
+		startExp, stopExp,
+		startBuffSize, stopBuffSize,
+	}: {
+			templateStart: string; templateStop: string;
+			startExp?: RegExp; stopExp?: RegExp;
+			startBuffSize?: number; stopBuffSize?: number;
+	}) {
+		this.templateStart = templateStart;
+		this.templateStop = templateStop;
+		this.startExp = startExp;
+		this.stopExp = stopExp;
+		this.startBuffSize = startBuffSize ?? this.templateStart.length;
+		this.stopBuffSize = stopBuffSize ?? this.templateStop.length;
+	}
+
+
+	/**
+	 * Returns the length of a match for this in [searchText], matching either
+	 * [template] or [regex], starting at [startIndex] or ending at [endIndex].
+	 * Note that the [regex], if provided, must have the global flag enabled.
+	 */
+	private matchLen({
+		searchText, template, startIndex, endIndex, regex,
+	}: {
+		searchText: string; template: string;
+		startIndex: number; endIndex: number; regex?: RegExp;
+	}): number {
```

Comment:
So matchLen is defined this way, with the weird double-definition, and matchStart is defined with a regular list of arguments? Shouldn't there be some consistency? Just a list of arguments should be fine if the defined object is never reused.

Reply:
I wanted to be able to pass an Object as an initializer to `RegionSpec.ts`.

I find
```ts
const changes = toggleGlobalSelectionFormat(view.state, 'Emphasis', new RegionSpec({
	// Template start/end
	templateStart: '_',
	templateStop: '_',

	// Regular expressions that match all possible start/ends
	startExp: /[_*]/g,
	stopExp: /[_*]/g,
}));
```
to be much more readable than
```ts
const changes = toggleGlobalSelectionFormat(view.state, 'Emphasis', new RegionSpec(
	// Template start/end
	'_',
	'_',

	// Regular expressions that match all possible start/ends
	/[_*]/g,
	/[_*]/g,
));
```

Of course, `RegionSpec` should really be an `interface` to avoid this double-declaration issue.

Reply:
I'd prefer we avoid inline interface definitions in general. A named interface is self documenting - you know immediately what it is. With inline, you have to look at a bunch of random parameters and try to see what this is about. Also if an interface is used only once I'm not certain we even need one, which is why regular arguments make more sense.

The bottom code above is fine even if slightly less readable - anyway the text editor can tell you what the arguments are. And if you push this logic, and the above example really is better, then the whole code base should be written like this, which is of course a much bigger discussion.